lib/remote_node.go

```diff
@@ -0,0 +1,599 @@
+	versionNonceReceived uint64
+	// userAgent is a meta level label that can be used to analyze the network.
+	userAgent string
+	// ServiceFlag is a bitfield that indicates the services supported by the peer.
+	ServiceFlag ServiceFlag
+	// StartingBlockHeight is the block height of the peer's block tip during the Version exchange.
+	StartingBlockHeight uint32
+	// minTxFeeRateNanosPerKB is the minimum transaction fee rate in nanos per KB that the peer will accept.
+	minTxFeeRateNanosPerKB uint64
+	// advertisedProtocolVersion is the protocol version advertised by the peer.
+	advertisedProtocolVersion ProtocolVersionType
+	// negotiatedProtocolVersion is the protocol version negotiated between the peer and our node. This is the minimum
+	// of the advertised protocol version and our node's protocol version.
+	negotiatedProtocolVersion ProtocolVersionType
+	// timeConnected is the unix timestamp of the peer, measured when the peer sent their Version message.
+	timeConnected *time.Time
+	// versionNegotiated is true if the peer passed the version negotiation step.
+	versionNegotiated bool
+	// timeOffsetSecs is the time offset between our node and the peer, measured by taking the difference between the
+	// peer's unix timestamp and our node's unix timestamp.
+	timeOffsetSecs int64
+
+	// ### The following fields are populated during the MsgDeSoVerack exchange.
+	// validatorPublicKey is the BLS public key of the peer, if the peer is a validator node.
+	validatorPublicKey *bls.PublicKey
+}
+
+func NewRemoteNode(id RemoteNodeId, srv *Server, bc *Blockchain, cmgr *ConnectionManager, params *DeSoParams,
+	minTxFeeRateNanosPerKB uint64, hyperSync bool, archivalMode bool, posValidator bool) *RemoteNode {
+	return &RemoteNode{
+		id:                     id,
+		connectionStatus:       RemoteNodeStatus_NotConnected,
+		handshakeMetadata:      nil,
+		srv:                    srv,
+		bc:                     bc,
+		cmgr:                   cmgr,
+		params:                 params,
+		minTxFeeRateNanosPerKB: minTxFeeRateNanosPerKB,
+		hyperSync:              hyperSync,
+		archivalMode:           archivalMode,
+		posValidator:           posValidator,
+	}
+}
+
+// setStatusValidated sets the connection status of the remote node to validated.
+func (rn *RemoteNode) setStatusValidated() {
+	rn.connectionStatus = RemoteNodeStatus_Validated
+}
+
+// setStatusConnected sets the connection status of the remote node to connected.
+func (rn *RemoteNode) setStatusConnected() {
+	rn.connectionStatus = RemoteNodeStatus_Connected
+}
+
+// setStatusTerminated sets the connection status of the remote node to terminated.
+func (rn *RemoteNode) setStatusTerminated() {
+	rn.connectionStatus = RemoteNodeStatus_Terminated
+}
+
+// setStatusAttempted sets the connection status of the remote node to attempted.
+func (rn *RemoteNode) setStatusAttempted() {
+	rn.connectionStatus = RemoteNodeStatus_Attempted
```

Comment:
Should these `setStatus` functions validate the state transition? For example, should `setStatusAttempted` complain if the status is `RemoteNodeStatus_Terminated`?

Reply:
I think it ends up cleaner if the caller should make this decision